### ISSUE TITLE
Type narrowing is not correctly propagated across functions

### ISSUE THREAD

**Issue by TalviT** [microsoft/pyright#11094]
**Description**
The type checker does not recognize that a variable has been checked for type constraints before it was passed to a function that calls another function that requires these constraints.

**Expected behavior**
There should be no error.

**Example code with error message**

```python
import os

ENV_VAR = os.getenv("SOME_ENV")  # Returns str | None

if ENV_VAR is None:
    raise ValueError

def second(var: str):
    print(var)

def first():
    second(ENV_VAR)
#          ^^^^^^^
#  Argument of type "str | None" cannot be assigned to parameter "var" of type "str" in #  function "second"
#   Type "str | None" is not assignable to type "str"
#     "None" is not assignable to "str"

first()
```

I'm using Pylance as VS Code extension. Version: 2025.9.1

**Comment by erictraut** [comment]
Pyright is working as intended here, so this isn't a bug.

Type narrowing is performed only within the context of an "execution scope" — a block of code with a code flow graph. Functions and methods have independent execution scopes and can be invoked in any order, so type narrowing in one execution scope cannot generally be applied to functions safely. Pyright does preserve type narrowing in specific cases where there is provably no way for the value to change. However, this can't be safely proven for module-scoped variables as in your example. The function `first` could be invoked by another module.
